File: 2주차/이라은/전력망을 둘로 나누기.py
from collections import deque

def solution(n, wires):
    answer = n
    q = deque()
    graph = []
    
    for _ in range(n+1):
        row = []
        graph.append(row)       
    
    # 트리 그래프 생성
    for wire in wires:
        v1, v2 = wire[0], wire[1]
        graph[v1].append(v2)
        graph[v2].append(v1)
    print(graph)
    
    def bfs(start, visited, graph):
        q = deque([start])
        cnt = 1
        visited[start] = True
        while q:
            now = q.popleft() #지금 탐색하는 지점
            for node in graph[now]:
                if not visited[node]:
                    visited[node] = True
                    q.append(node)
                    cnt += 1
        return cnt
                        
    # wire가 하나씩 없다고 가정
    for start, end in wires:
        visited = [False] * (n+1) 
        visited[end] = True # 방문처리 배열을 활용하여 매 반복마다 방문처리 한다
        result = bfs(start, visited, graph)    
        answer = min(answer, abs(result - (n-result)))
    
    return answer

File: 2주차/김태경/전력망을 둘로 나누기.java
import java.util.*;
class Solution {
    ArrayList<Integer>[] graph;
    boolean visited1[];
    boolean visited2[];
    public int solution(int n, int[][] wires) {
        int answer = 101;
        this.visited1=new boolean[n+1];
        this.visited2=new boolean[n+1];
        this.graph=new ArrayList[n+1];
        for(int i=1;i<n+1;i++){
            graph[i]=new ArrayList<>();
        }
        for(int[] wire : wires){
            int i=wire[0];
            int j=wire[1];
            graph[i].add(j);
            graph[j].add(i);
        }
        for(int i=1;i<n+1;i++){
            for(int j : graph[i]){
                int a = bfs(i,j);
                int b=bfs(j,i);
                //answer=Math.min(Math.abs(a-(n-a)),answer);
                answer=Math.min(Math.abs(a-b),answer);
                //System.out.println(answer+ " a="+a+" b="+b);
                visited1=new boolean[n+1];
                // visited1[j]=false;
                // visited1[i]=false;
            }
        }
        
        
        return answer;
    }
    
    int bfs(int s, int d){
        visited1[s]=true;
        visited1[d]=true;
        int cnt=1;
        ArrayDeque<Integer> dq = new ArrayDeque<>();
        dq.offer(s);
        while(!dq.isEmpty()){
            int now = dq.poll();
            for(int k =0; k<graph[now].size();k++){
                if(!visited1[graph[now].get(k)]){
                    cnt++;
                    visited1[graph[now].get(k)]=true;
                    dq.offer(graph[now].get(k));
                        
                }
            }
            
        }
        return cnt;
    }
}
